File: src/main/java/nl/mpcjanssen/simpletask/CalendarSync.java
/**
 *
 * Copyright (c) 2015 Vojtech Kral
 *
 * LICENSE:
 *
 * Simpletas is free software: you can redistribute it and/or modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any
 * later version.
 *
 * Simpletask is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with Sinpletask.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 * @author Vojtech Kral
 * @license http://www.gnu.org/licenses/gpl.html
 * @copyright 2015 Vojtech Kral
 */

package nl.mpcjanssen.simpletask;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.database.Cursor;
import android.graphics.Color;
import android.net.Uri;
import android.provider.CalendarContract;
import android.provider.CalendarContract.Calendars;
import android.provider.CalendarContract.Events;
import android.provider.CalendarContract.Reminders;
import hirondelle.date4j.DateTime;
import nl.mpcjanssen.simpletask.task.Task;
import nl.mpcjanssen.simpletask.task.TodoList;
import nl.mpcjanssen.simpletask.util.Util;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Calendar;
import java.util.List;
import java.util.TimeZone;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;


public class CalendarSync {
    private static final String TAG = CalendarSync.class.getSimpleName();
    private static final String PACKAGE = TodoApplication.getAppContext().getPackageName();
    private static final TimeZone UTC = TimeZone.getTimeZone("UTC");

    private static final String ACCOUNT_NAME = "Simpletask Calendar";
    private static final String ACCOUNT_TYPE = CalendarContract.ACCOUNT_TYPE_LOCAL;
    private static final Uri CAL_URI = Calendars.CONTENT_URI.buildUpon()
            .appendQueryParameter(CalendarContract.CALLER_IS_SYNCADAPTER, "true")
            .appendQueryParameter(Calendars.ACCOUNT_NAME, ACCOUNT_NAME)
            .appendQueryParameter(Calendars.ACCOUNT_TYPE, ACCOUNT_TYPE).build();
    private static final String CAL_NAME = "simpletask_reminders_v34SsjC7mwK9WSVI";
    private static final int CAL_COLOR = Color.BLUE;       // Chosen arbitrarily...
    private static final int EVT_DURATION_DAY = 24*60*60*1000;  // ie. 24 hours

    private static final int SYNC_DELAY_MS = 1000;
    private final Logger log;

    private class SyncRunnable implements Runnable {
        @Override
        public void run() {
            try {
                CalendarSync.this.sync();
            } catch (Exception e) {
                log.error("STPE exception", e);
            }
        }
    }

    private TodoApplication m_app;
    private SyncRunnable m_sync_runnable;
    private ContentResolver m_cr;
    private int m_sync_type;
    private int m_rem_margin = 1440;
    private DateTime m_rem_time = DateTime.forTimeOnly(12, 0, 0, 0);
    private ScheduledThreadPoolExecutor m_stpe;

    private long getCalID() {
        final String[] projection = {Calendars._ID, Calendars.NAME};
        final String selection = Calendars.NAME+" = ?";
        final String[] args = {CAL_NAME};
        Cursor cursor = m_cr.query(CAL_URI, projection, selection, args, null);
        if (cursor == null) return -1;
        if (cursor.getCount() == 0) return -1;
        cursor.moveToFirst();
        long ret = cursor.getLong(0);
        cursor.close();
        return ret;
    }

    private void addCalendar(boolean warnCalExists) {
        if (getCalID() != -1) {
            if (warnCalExists) {
                log.warn("Calendar already exists, overwriting...");
                Util.showToastShort(TodoApplication.getAppContext(), R.string.calendar_exists_warning);
            }
            return;
        }

        final ContentValues cv = new ContentValues();
        cv.put(Calendars.ACCOUNT_NAME, ACCOUNT_NAME);
        cv.put(Calendars.ACCOUNT_TYPE, ACCOUNT_TYPE);
        cv.put(Calendars.NAME, CAL_NAME);
        cv.put(Calendars.CALENDAR_DISPLAY_NAME, m_app.getString(R.string.calendar_disp_name));
        cv.put(Calendars.CALENDAR_COLOR, CAL_COLOR);
        cv.put(Calendars.CALENDAR_ACCESS_LEVEL, Calendars.CAL_ACCESS_READ);
        cv.put(Calendars.OWNER_ACCOUNT, ACCOUNT_NAME);
        cv.put(Calendars.VISIBLE, 1);
        cv.put(Calendars.SYNC_EVENTS, 1);

        Uri calUri = m_cr.insert(CAL_URI, cv);
        boolean added = true;
        if (calUri == null) added = false;
        else if (getCalID() == -1) added = false;    // This might happen eg. because of CM's privacy guard

        if (added) {
            log.debug("Calendar added");
        }
        else {
            log.error("Could not add calendar");
            m_sync_type = 0;
        }
    }

    private void removeCalendar() {
        final String selection = Calendars.NAME+" = ?";
        final String[] args = {CAL_NAME};
        int ret = m_cr.delete(CAL_URI, selection, args);
        if (ret == 1) log.debug("Calendar removed");
        else log.warn("Unexpected return value while removing calendar: " + ret);
    }

    @SuppressWarnings("NewApi")
    private void insertEvt(long calID, DateTime date, String title, String description) {
        ContentValues values = new ContentValues();

        TimeZone localZone = Calendar.getInstance().getTimeZone();
        long dtstart = date.getMilliseconds(UTC);

        // Event:
        values.put(Events.CALENDAR_ID, calID);
        values.put(Events.TITLE, title);
        values.put(Events.DTSTART, dtstart);
        values.put(Events.DTEND, dtstart + EVT_DURATION_DAY);  // Needs to be set to DTSTART +24h, otherwise reminders don't work
        values.put(Events.ALL_DAY, 1);
        values.put(Events.DESCRIPTION, description);
        values.put(Events.EVENT_TIMEZONE, UTC.getID());
        values.put(Events.STATUS, Events.STATUS_CONFIRMED);
        values.put(Events.HAS_ATTENDEE_DATA, true);      // If this is not set, Calendar app is confused about Event.STATUS
        values.put(Events.CUSTOM_APP_PACKAGE, PACKAGE);
        values.put(Events.CUSTOM_APP_URI, Uri.withAppendedPath(Simpletask.URI_SEARCH, title).toString());
        Uri uri = m_cr.insert(Events.CONTENT_URI, values);

        // Reminder:
        // Only create reminder if it's in the future, otherwise it would go off immediately
        // NOTE: DateTime.minus()/plus() only accept values >=0 and <10000 (goddamnit date4j!), hence the division.
        DateTime remDate = date.minus(0, 0, 0,  m_rem_margin / 60,  m_rem_margin % 60, 0, 0, DateTime.DayOverflow.Spillover);
        remDate = remDate.plus(0, 0, 0, m_rem_time.getHour(), m_rem_time.getMinute(), 0, 0, DateTime.DayOverflow.Spillover);
        if (remDate.isInTheFuture(localZone)) {
            long evtID = Long.parseLong(uri.getLastPathSegment());
            values.clear();
            values.put(Reminders.EVENT_ID, evtID);
            values.put(Reminders.MINUTES, remDate.numSecondsFrom(date) / 60);
            values.put(Reminders.METHOD, Reminders.METHOD_ALERT);
            m_cr.insert(Reminders.CONTENT_URI, values);
        }
    }

    private void insertEvts(long calID, final List<Task> tasks) {
        for (Task task: tasks) {
            if (task.isCompleted()) continue;

            DateTime dt;
            String text = null;

            // Check due date:
            if ((m_sync_type & SYNC_TYPE_DUES) != 0) {
                dt = task.getDueDate();
                if (dt != null) {
                    text = task.getText();
                    insertEvt(calID, dt, text, m_app.getString(R.string.calendar_sync_desc_due));
                }
            }

            // Check threshold date:
            if ((m_sync_type & SYNC_TYPE_THRESHOLDS) != 0) {
                dt = task.getThresholdDate();
                if (dt != null) {
                    if (text == null) text = task.getText();
                    insertEvt(calID, dt, text, m_app.getString(R.string.calendar_sync_desc_thre));
                }
            }
        }
    }

    private void purgeEvts(long calID) {
        final String selection = Events.CALENDAR_ID+" = ?";
        final String args[] = {""};
        args[0] = String.valueOf(calID);
        m_cr.delete(Events.CONTENT_URI, selection, args);
    }

    private void sync() {
        if (m_sync_type == 0) return;

        TodoList tl =  m_app.getTodoList();

        final List<Task> tasks =tl.getTasks();
        setReminderDays(m_app.getReminderDays());
        setReminderTime(m_app.getReminderTime());

        long calID = getCalID();
        if (calID == -1) {
            log.error("sync(): No calendar!");
            return;
        }

        log.debug("Syncing due/threshold calendar reminders...");
        purgeEvts(calID);
        insertEvts(calID, tasks);
    }

    private void setSyncType(int syncType, boolean warnCalExists) {
        if (syncType == m_sync_type) return;
        if (!TodoApplication.ATLEAST_API16) {
            m_sync_type = 0;
            return;
        }
        int old_sync_type = m_sync_type;
        m_sync_type = syncType;

        if ((old_sync_type == 0) && (m_sync_type > 0)) {
            addCalendar(warnCalExists);
        }
        else if ((old_sync_type > 0) && (m_sync_type == 0)) {
            removeCalendar();
        }

        if (m_sync_type > 0) syncLater();
    }



    public static final int SYNC_TYPE_DUES = 1, SYNC_TYPE_THRESHOLDS = 2;

    public CalendarSync(TodoApplication app, boolean syncDues, boolean syncThresholds) {
        log = LoggerFactory.getLogger(this.getClass());
        m_app = app;
        m_sync_runnable = new SyncRunnable();
        m_cr = app.getContentResolver();
        m_stpe = new ScheduledThreadPoolExecutor(1);
        int syncType = 0;
        if (syncDues) syncType |= SYNC_TYPE_DUES;
        if (syncThresholds) syncType |= SYNC_TYPE_THRESHOLDS;
        setSyncType(syncType, false);
    }

    public void syncLater() {
        if (m_sync_type == 0) return;

        m_stpe.getQueue().clear();
        m_stpe.schedule(m_sync_runnable, SYNC_DELAY_MS, TimeUnit.MILLISECONDS);
    }

    public void setSyncDues(boolean bool) {
        int syncType = bool ? m_sync_type | SYNC_TYPE_DUES : m_sync_type & ~SYNC_TYPE_DUES;
        setSyncType(syncType, true);
    }

    public void setSyncThresholds(boolean bool) {
        int syncType = bool ? m_sync_type | SYNC_TYPE_THRESHOLDS : m_sync_type & ~SYNC_TYPE_THRESHOLDS;
        setSyncType(syncType, true);
    }

    public void setReminderDays(int days) {
        m_rem_margin = days * 1440;
    }

    public void setReminderTime(int time) {
        m_rem_time = DateTime.forTimeOnly(time / 60, time % 60, 0, 0);
    }
}

File: app/src/main/java/nl/mpcjanssen/simpletask/CalendarSync.kt
/**

 * Copyright (c) 2015 Vojtech Kral

 * LICENSE:

 * Simpletask is free software: you can redistribute it and/or modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any
 * later version.

 * Simpletask is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.

 * You should have received a copy of the GNU General Public License along with Sinpletask.  If not, see
 * //www.gnu.org/licenses/>.

 * @author Vojtech Kral
 * *
 * @license http://www.gnu.org/licenses/gpl.html
 * *
 * @copyright 2015 Vojtech Kral
 */

package nl.mpcjanssen.simpletask

import java.util.*
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit
import android.Manifest
import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.ContentResolver
import android.content.ContentValues
import android.content.ContentProviderOperation
import android.content.pm.PackageManager
import android.database.Cursor
import android.graphics.Color
import android.net.Uri
import android.provider.CalendarContract.*
import androidx.core.content.ContextCompat
import android.util.Log
import hirondelle.date4j.DateTime
import nl.mpcjanssen.simpletask.task.*
import nl.mpcjanssen.simpletask.util.toDateTime

private enum class EvtStatus {
    KEEP,
    DELETE,
    INSERT,
}

data class EvtKey(val dtStart: Long, val title: String)

private class Evt(
        var id: Long,
        var dtStart: Long,
        var title: String,
        var description: String,
        var remID: Long = -1,
        var remMinutes: Long = -1,
        var status: EvtStatus = EvtStatus.DELETE) {

    constructor(cursor: Cursor) // Create from a db record
            : this(cursor.getLong(0), cursor.getLong(2), cursor.getString(3), cursor.getString(4))

    constructor(date: DateTime, title: String, desc: String) // Create from a Task
            : this(-1, date.getMilliseconds(CalendarSync.UTC), title, desc, -1, -1, EvtStatus.INSERT) {

        val localZone = Calendar.getInstance().timeZone
        val remMargin = TodoApplication.config.reminderDays * 1440
        val remTime = TodoApplication.config.reminderTime
        val remDT = DateTime.forTimeOnly(remTime / 60, remTime % 60, 0, 0)

        // Reminder data:
        // Only create reminder if it's in the future, otherwise it would go off immediately
        // NOTE: DateTime.minus()/plus() only accept values >=0 and <10000 (goddamnit date4j!), hence the division.
        var remDate = date.minus(0, 0, 0, remMargin / 60, remMargin % 60, 0, 0, DateTime.DayOverflow.Spillover)
        remDate = remDate.plus(0, 0, 0, remDT.hour, remDT.minute, 0, 0, DateTime.DayOverflow.Spillover)
        if (remDate.isInTheFuture(localZone)) {
            remID = 0L // 0 = reminder entry to be created
            remMinutes = remDate.numSecondsFrom(date) / 60
        }
    }

    fun key() = EvtKey(dtStart, title)

    override fun hashCode(): Int {
        var result = dtStart.hashCode()
        result = 31 * result + title.hashCode()
        result = 31 * result + description.hashCode()
        result = 31 * result + remMinutes.hashCode()
        return result
    }

    override fun equals(other: Any?) =
            if (other !is Evt) {
                false
            }  else {
            dtStart == other.dtStart &&
                    title == other.title &&
                    description == other.description &&
                    remMinutes == other.remMinutes }

    @TargetApi(16)
    fun addOp(list: ArrayList<ContentProviderOperation>, calID: Long) {
        if (status == EvtStatus.DELETE) {
            val args = arrayOf(id.toString())
            list.add(ContentProviderOperation.newDelete(Events.CONTENT_URI)
                    .withSelection(Events._ID + "=?", args)
                    .build())
            if (remID >= 0) {
                val remArgs = arrayOf(remID.toString())
                list.add(ContentProviderOperation.newDelete(Reminders.CONTENT_URI)
                        .withSelection(Reminders._ID + "=?", remArgs)
                        .build())
            }
        } else if (status == EvtStatus.INSERT) {
            list.add(ContentProviderOperation.newInsert(Events.CONTENT_URI)
                    .withValue(Events.CALENDAR_ID, calID)
                    .withValue(Events.TITLE, title)
                    .withValue(Events.DTSTART, dtStart)
                    .withValue(Events.DTEND, dtStart + 24 * 60 * 60 * 1000) // Needs to be set to DTSTART +24h, otherwise reminders don't work
                    .withValue(Events.ALL_DAY, 1)
                    .withValue(Events.DESCRIPTION, description)
                    .withValue(Events.EVENT_TIMEZONE, CalendarSync.UTC.id)
                    .withValue(Events.STATUS, Events.STATUS_CONFIRMED)
                    .withValue(Events.HAS_ATTENDEE_DATA, true) // If this is not set, Calendar app is confused about Event.STATUS
                    .withValue(Events.CUSTOM_APP_PACKAGE, TodoApplication.app.packageName)
                    .withValue(Events.CUSTOM_APP_URI, Uri.withAppendedPath(Simpletask.URI_SEARCH, title).toString())
                    .build())
            if (remID >= 0) {
                val evtIdx = list.size - 1
                list.add(ContentProviderOperation.newInsert(Reminders.CONTENT_URI)
                        .withValueBackReference(Reminders.EVENT_ID, evtIdx)
                        .withValue(Reminders.MINUTES, remMinutes)
                        .withValue(Reminders.METHOD, Reminders.METHOD_ALERT)
                        .build())
            }
        }
    }
}

private class SyncStats(val inserts: Long, val keeps: Long, val deletes: Long)

/**
 * A hashmap of Evts
 *
 * In order to make calendar notifications sync more efficient, we want to only add/remove those
 * that actually need to be added/removed. However, Simpletask doesn't track task identity
 * and consequently we don't know what changed, not to mention todo.txt might've also
 * been changed outside of Simpletask.
 *
 * And so we need to diff between task list and what's actually in the calendar.
 * To do that, this hashmap is used, which stores events (Evt) by their date and text (EvtKey).
 * (Note that there might be multiple equal events, hence the LinkedList as the map value type.)
 * When constructed, the hashmap loads events from the DB, all of which are initially marked for deletion.
 * Then, task list is merged in: Each event (that comes from a task) is located in the hashmap
 * - if an equal one is found, it is marked to be kept unchanged.
 * If not, new event is inserted into the map and marked for insertion.
 *
 * Finally, the hashmap contents are applied - contained events are iterated and inserted/deleted as appropriate.
 * This is done using ContentResolver.applyBatch for better efficiency.
 */
@SuppressLint("Recycle", "NewAPI")
private class EvtMap private constructor() : HashMap<EvtKey, LinkedList<Evt>>() {
    @SuppressLint("MissingPermission")
    constructor(cr: ContentResolver, calID: Long) : this() {
        val evtPrj = arrayOf(Events._ID, Events.CALENDAR_ID, Events.DTSTART, Events.TITLE, Events.DESCRIPTION)
        val evtSel = "${Events.CALENDAR_ID} = ?"
        val evtArgs = arrayOf(calID.toString())

        val remPrj = arrayOf(Reminders._ID, Reminders.EVENT_ID, Reminders.MINUTES)
        val remSel = "${Reminders.EVENT_ID} = ?"

        val evts = cr.query(Events.CONTENT_URI, evtPrj, evtSel, evtArgs, null)
                ?: throw IllegalArgumentException("null cursor")
        while (evts.moveToNext()) {
            val evt = Evt(evts)

            // Try to find a matching reminder
            val remArgs = arrayOf(evt.id.toString())
            val rem = cr.query(Reminders.CONTENT_URI, remPrj, remSel, remArgs, null)
                    ?: throw IllegalArgumentException("null cursor")
            if (rem.count > 0) {
                rem.moveToFirst()
                evt.remID = rem.getLong(0)
                evt.remMinutes = rem.getLong(2)
            }
            rem.close()

            // Insert into the hashmap
            val evtkey = evt.key()
            var list = this.get(evtkey)
            if (list != null) list.add(evt)
            else {
                list = LinkedList<Evt>()
                list.add(evt)
                this.put(evtkey, list)
            }
        }
        evts.close()

    }

    fun mergeEvt(evt: Evt) {
        val key = evt.key()
        val list = this.get(key)
        evt.status = EvtStatus.INSERT
        if (list == null) {
            val nlist = LinkedList<Evt>()
            nlist.add(evt)
            this.put(key, nlist)
        } else {
            for (oevt in list) {
                if (oevt.status == EvtStatus.DELETE && oevt == evt) {
                    oevt.status = EvtStatus.KEEP
                    return
                }
            }
            list.add(evt)
        }
    }

    fun mergeTask(task: Task) {

            if (task.isCompleted()) return

            var text: String? = null

            // Check due date:
            var dt = task.dueDate?.toDateTime()
            if (TodoApplication.config.isSyncDues && dt != null) {
                text = task.showParts(CalendarSync.TASK_TOKENS)
                val evt = Evt(dt, text, TodoApplication.app.getString(R.string.calendar_sync_desc_due))
                mergeEvt(evt)
            }

            // Check threshold date:
            dt = task.thresholdDate?.toDateTime()
            if (TodoApplication.config.isSyncThresholds && dt != null) {
                if (text == null) text = task.showParts(CalendarSync.TASK_TOKENS)
                val evt = Evt(dt, text, TodoApplication.app.getString(R.string.calendar_sync_desc_thre))
                mergeEvt(evt)
            }
    }

    @SuppressLint("NewApi")
    fun apply(cr: ContentResolver, calID: Long): SyncStats {
        val ops = ArrayList<ContentProviderOperation>()
        var ins = 0L
        var kps = 0L
        var dels = 0L

        for (list in values) {
            for (evt in list) {
                when {
                    evt.status == EvtStatus.INSERT -> ins++
                    evt.status == EvtStatus.KEEP -> kps++
                    evt.status == EvtStatus.DELETE -> dels++
                }

                evt.addOp(ops, calID)
            }
        }

        cr.applyBatch(AUTHORITY, ops)
        return SyncStats(ins, kps, dels)
    }
}

object CalendarSync {

    private val ACCOUNT_NAME = "Simpletask Calendar"
    private val ACCOUNT_TYPE = ACCOUNT_TYPE_LOCAL
    private val CAL_URI = Calendars.CONTENT_URI.buildUpon()
            .appendQueryParameter(CALLER_IS_SYNCADAPTER, "true")
            .appendQueryParameter(Calendars.ACCOUNT_NAME, ACCOUNT_NAME)
            .appendQueryParameter(Calendars.ACCOUNT_TYPE, ACCOUNT_TYPE)
            .build()
    private val CAL_NAME = "simpletask_reminders_v34SsjC7mwK9WSVI"
    private val CAL_COLOR = Color.BLUE // Chosen arbitrarily...

    private val SYNC_DELAY_MS = 20 * 1000
    private val TAG = "CalendarSync"

    val UTC: TimeZone = TimeZone.getTimeZone("UTC")
    val TASK_TOKENS: (TToken) -> Boolean = {
        when (it) {
            is CompletedToken, is CreateDateToken, is CompletedDateToken, is PriorityToken, is ThresholdDateToken,
            is DueDateToken, is HiddenToken, is RecurrenceToken -> false
            else -> true
        }
    }

    private class SyncRunnable : Runnable {
        override fun run() {
            try {
                sync()
            } catch (e: Exception) {
                Log.e(TAG, "STPE exception", e)
            }

        }
    }

    private val m_sync_runnable: SyncRunnable
    private val m_cr: ContentResolver
    private val m_stpe: ScheduledThreadPoolExecutor

    @SuppressLint("Recycle")
    private fun findCalendar(): Long {
        val projection = arrayOf(Calendars._ID, Calendars.NAME)
        val selection = Calendars.NAME + " = ?"
        val args = arrayOf(CAL_NAME)
        /* Check for calendar permission */
        val permissionCheck = ContextCompat.checkSelfPermission(TodoApplication.app,
                Manifest.permission.WRITE_CALENDAR)

        if (permissionCheck == PackageManager.PERMISSION_DENIED) {
            if (TodoApplication.config.isSyncDues || TodoApplication.config.isSyncThresholds) {
                throw IllegalStateException("no calendar access")
            } else {
                return -1
            }
        }

        val cursor = m_cr.query(CAL_URI, projection, selection, args, null)
                ?: throw IllegalArgumentException("null cursor")
        if (cursor.count == 0) {
            cursor.close()
            return -1
        }
        cursor.moveToFirst()
        val ret = cursor.getLong(0)
        cursor.close()
        return ret
    }

    private fun addCalendar() {
        val cv = ContentValues()
        cv.apply {
            put(Calendars.ACCOUNT_NAME, ACCOUNT_NAME)
            put(Calendars.ACCOUNT_TYPE, ACCOUNT_TYPE)
            put(Calendars.NAME, CAL_NAME)
            put(Calendars.CALENDAR_DISPLAY_NAME, TodoApplication.app.getString(R.string.calendar_disp_name))
            put(Calendars.CALENDAR_COLOR, CAL_COLOR)
            put(Calendars.CALENDAR_ACCESS_LEVEL, Calendars.CAL_ACCESS_READ)
            put(Calendars.OWNER_ACCOUNT, ACCOUNT_NAME)
            put(Calendars.VISIBLE, 1)
            put(Calendars.SYNC_EVENTS, 1)
        }
        m_cr.insert(CAL_URI, cv)
    }

    @SuppressLint("NewApi")
    private fun removeCalendar() {
        Log.d(TAG, "Removing Simpletask calendar")
        val selection = Calendars.NAME + " = ?"
        val args = arrayOf(CAL_NAME)
        try {
            val ret = m_cr.delete(CAL_URI, selection, args)
            when (ret) {
                0 -> Log.d(TAG, "No calendar to remove")
                1 -> Log.d(TAG, "Calendar removed")
                else -> Log.e(TAG, "Unexpected return value while removing calendar: " + ret)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error while removing calendar", e)
        }
    }

    private fun sync() {
        Log.d(TAG, "Checking whether calendar sync is needed")
        try {
            var calID = findCalendar()

            if (!TodoApplication.config.isSyncThresholds && !TodoApplication.config.isSyncDues) {
                if (calID >= 0) {
                    Log.d(TAG, "Calendar sync not enabled")
                    removeCalendar()
                }
                return
            }

            if (calID < 0) {
                addCalendar()
                calID = findCalendar() // Re-find the calendar, this is needed to verify it has been added
                if (calID < 0) {
                    // This happens when CM privacy guard disallows to write calendar (1)
                    // OR it allows to write calendar but disallows reading it (2).
                    // Either way, we cannot continue, but before bailing,
                    // try to remove Calendar in case we're here because of (2).
                    Log.d(TAG, "No access to Simpletask calendar")
                    removeCalendar()
                    throw IllegalStateException("Calendar nor added")
                }
            }

            Log.d(TAG, "Syncing due/threshold calendar reminders...")
            val evtmap = EvtMap(m_cr, calID)
            TodoApplication.todoList.each {
                evtmap.mergeTask(it)
            }
            val stats = evtmap.apply(m_cr, calID)
            Log.d(TAG, "Sync finished: ${stats.inserts} inserted, ${stats.keeps} unchanged, ${stats.deletes} deleted")
        } catch (e: SecurityException) {

            Log.e(TAG, "No calendar access permissions granted", e )
        } catch (e: Exception) {
            Log.e(TAG, "Calendar error", e)
        }
    }

    fun updatedSyncTypes() {
        syncLater()
    }

    init {
        m_sync_runnable = SyncRunnable()
        m_cr = TodoApplication.app.contentResolver
        m_stpe = ScheduledThreadPoolExecutor(1)
    }

    fun syncLater() {
        m_stpe.queue.clear()
        m_stpe.schedule(m_sync_runnable, SYNC_DELAY_MS.toLong(), TimeUnit.MILLISECONDS)
    }
}

